### ISSUE TITLE
Can't select a mission, error: [string "[T] @libs/SpaceStation.lua"]:599: Attempt to lock ref * disallowed. Ref ** is already locked 1 times.

### ISSUE THREAD

**Issue by Echoes91** [pioneerspacesim/pioneer#4832]
### Observed behaviour

No dialog displayed when clicking on a mission from bulletin board. Game crashes when clicking on another mission:

```
error: [string "[T] @libs/SpaceStation.lua"]:599: Attempt to lock ref * disallowed. Ref ** is already locked 1 times.
stack traceback:
	[C]: in function 'assert'
	[string "[T] @libs/SpaceStation.lua"]:599: in function 'LockAdvert'
	[string "[T] @pigui/modules/station-view/02-bulletinBo..."]:145: in function 'onClickItem'
	[string "[T] @pigui/libs/table.lua"]:130: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:137: in function 'child'
	[string "[T] @pigui/libs/table.lua"]:81: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:167: in function 'withStyleVars'
	[string "[T] @pigui/libs/table.lua"]:80: in function 'render'
	[string "[T] @pigui/modules/station-view/02-bulletinBo..."]:159: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:167: in function 'withStyleVars'
	...
	[string "[T] @pigui/views/tab-view.lua"]:96: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:158: in function 'withStyleColors'
	[string "[T] @pigui/views/tab-view.lua"]:92: in function 'renderTabView'
	[string "[T] @pigui/views/sterror: Frame instance deletion outside 'DeleteFrame' [0]
```

### Expected behaviour

Dialog displayed when clicking on a mission from bulletin board

### Steps to reproduce

1. Start a game
2. Show comms - Bulletin board
4. Click on a mission
5. Click on another mission

__My pioneer version (and OS):__
pioneer 20200308 (4f005e44a) on Arch Linux x86_64

![PSS_error_1](https://user-images.githubusercontent.com/11097013/76172084-58228c00-6192-11ea-90db-d9aadd8030ee.png)
![PSS_error_2](https://user-images.githubusercontent.com/11097013/76172086-5953b900-6192-11ea-998d-eed9825e6782.png)

**Comment by impaktor** [comment]
Thanks for reporting. Possibly something @vakhoir might want to have a look at.

**Comment by pcercuei** [comment]
I usually get these errors when running a recently compiled Pioneer with old Lua files. Make sure the installed data is up-to-date.

**Comment by Echoes91** [comment]
Installation is a fresh master cloned yesterday, no modifications.

**_Update_** No issue with tag 20200203 built today, this narrows the filed to very recent changes

**Comment by pcercuei** [comment]
Try to rm -rf the old install, and install it again.

Note that even if that fixes it, it's still a bug to me, the game shouldn't load old leftover files.

**Comment by CylonicRaider** [comment]
Bisecting between tag `20200203` and current `master` (commit c2f3305730b3d0a08f6b31f86f494a10c8640a0a as of this writing) points at commit 4bce34b081775d74c33ae41e170aa2f28c3b18c9 as the culprit, which looks like it could be related indeed.

**Comment by Web-eWorks** [comment]
@CylonicRaider thank you for the bisect, and my apologies for taking so long to address this issue.
Judging by what the referenced commit touched, it looks like we may have been unlocking the ref in the newUI code and that didn't get properly transferred into the pigui version, causing issues once we struck the newUI code entirely.

I find it interesting that this is happening *at all*, given that I've tested the mentioned reproduction steps and could not click another mission without first closing the modal window, properly unlocking the reference and not encountering the bug.

I'm also not able to reproduce the mentioned error where the mission dialog isn't populated correctly; I'd appreciate it if @Echoes91 or @pcercuei could please double-check this is present on a fresh development environment (e.g. clone `pioneerspacesim/pioneer` to a new folder and follow the build steps there), and closely document the specific actions you took to produce the error.

Regarding the `Frame instance deletion outside DeleteFrame`, that's simply an artifact of the way we throw errors and can be reproduced with any error condition that `assert()`s the entire process. We'll track and/or fix that in a separate issue.

**Comment by Echoes91** [comment]
Fresh clone of today:

```
pioneer 20200317 (133c31d3d)


--------------------
SDL Version (build) 2.0.12
SDL Version (dynamic) 2.0.12
SDL Versions match
SDL_image Version (build): 2.0.5
SDL_image Version (dynamic): 2.0.5
SDL_image Versions match
Assimp Version: 5.0.0
FreeType Version: 2.10.1
GLEW dynamic version: 2.0.0
--------------------
```

### To reproduce:

1. Start new game
![image](https://user-images.githubusercontent.com/11097013/76865985-c686de80-6863-11ea-82f4-616c7bb2cc15.png)
 
2. Open "Show comms" (F4) - "Bulletin board"
![image](https://user-images.githubusercontent.com/11097013/76866223-19609600-6864-11ea-8fd0-64f7ee32c5dc.png)

3. Click on _any_ row -> _no dialog appears, maybe this was the modal window that should have blocked subsequent clicks?_
![image](https://user-images.githubusercontent.com/11097013/76866379-59277d80-6864-11ea-9a4a-d704a5f59e74.png)

4. Click on _any other_ row -> _error message appears_
![image](https://user-images.githubusercontent.com/11097013/76866523-8bd17600-6864-11ea-8b93-e64565994718.png)

Unfortunately i have zero experience with GUIs, from your comment I'd think that once the dialog is fixed so that it appears and closes properly (locking and unlocking the reference), then the error will be avoided.

**Comment by impaktor** [comment]
I can reproduce it as well.

**Comment by Web-eWorks** [comment]
@Echoes91 thank you very much for the detailed reproduction steps! I'll take a very close look at that and try to track down the underlying cause.

**Comment by vakhoir** [comment]
@Web-eWorks, if you haven't started yet, I'll take care of this.

**Comment by Web-eWorks** [comment]
@vakhoir I've managed to intermittently reproduce the issue, but I've not had time to work on pioneer for the last few days due to some family commitments. Feel free to take over!

**Comment by vakhoir** [comment]
I narrowed it down to 4bce34b081775d74c33ae41e170aa2f28c3b18c9. 

This is currently speculation, but I think it's caused by moving the StationView to the new PiGuiView class, and modals being drawn in handlers added through registerModule. IMGui is very about being in the right window context when drawing a modal, and I'm guessing that by the time the 'game' module handler is called (see bottom of /data/pigui/libs/modal-win.lua), we've already jumped out of the window where we actually want to draw it.

```
// this class should be able to
// dispatch to lua-registered handler functions and call them during the Draw3D
// method.
```

Does "should be" here mean it's a future aspiration, or something we can currently do? Is there a working example somewhere? I think adding the modal drawing to the StationView handler should do the trick.

**Comment by Web-eWorks** [comment]
@vakhoir
> This is currently speculation, but I think it's caused by moving the StationView to the new PiGuiView class, and modals being drawn in handlers added through registerModule. IMGui is very about being in the right window context when drawing a modal, and I'm guessing that by the time the 'game' module handler is called (see bottom of /data/pigui/libs/modal-win.lua), we've already jumped out of the window where we actually want to draw it.

The only thing that replacing the UIView with a PiGuiView should affect is that the newUI code isn't being executed anymore. Nothing in the UIView class triggered or affected execution of pigui code to the best of my knowledge. I've postulated a little above, but I believe there's a piece of code (or more probably, a side-effect of that code) in the newUI version of the StationView that didn't get re-implemented in the pigui version.

> Does "should be" here mean it's a future aspiration, or something we can currently do? Is there a working example somewhere? I think adding the modal drawing to the StationView handler should do the trick.

"Should be" means "in the future" in this case. If you look at `Pi.cpp` in the MainLoop method, you'll see that we start the ImGui frame *after* we've called `m_currentView->Draw`, due to some unavoidable ordering issues with input handling, etc. This logically means we can't call pigui stuff from PiGuiView right now, because there's no ImGui state set up to record those calls properly.

One of the items on my backlog is to reorder the main loop so that we start the ImGui frame at the very beginning of the loop to allow *all* game code to run ImGui methods; whether that's calling into Lua so pigui can draw the UI, or to make immediate calls for the purpose of debugging logic / data.

**Comment by vakhoir** [comment]
> The only thing that replacing the UIView with a PiGuiView should affect is that the newUI code isn't being executed anymore. 

Yeah, I think I figured out how that could lead to the issue. The Info-View, Station-View and the ModalWindow manager are all modules registered with `pigui.registerModule`. The modal window manager seems to be the first one registered, and the modules are run in the ordered they were registered, that means the Station-View is render on top of the modal window, and thus not visible. I'm guessing that by rendering the newUI view we somehow end up letting the modal window be rendered as well. I don't feel like investigating how exactly *that* happens, but I confirmed that prepending a module in `pigui.registerModule` instead of appending it fixes the issue. That could be a quick and dirty fix for anyone who wants to get their local build working.

As for how to actually fix the issue, I see 2 possibilities:

- Remove the Modal Window manager from modules and have them triggered from the Station/Info Views
- Add a way to set a priority for a module, when you register it, and set the Modal Window manager to the lowest priority.

I'm leaning towards the second option. The whole appeal of the Modal Window manager class is that it takes care of everything for you, and you can create a modal from any window you want.

Thoughts?

**Comment by Web-eWorks** [comment]
@vakhoir If I'm reading what you've written correctly, I've implemented a possible solution in #4842 - simply adding a separate module group `'modal'` that gets run at the end of the frame. *If* I understand the dilemma correctly, I think that should resolve the behavior you're describing.

**Comment by vakhoir** [comment]
Yeah, that should do it. I'll test and make any necessary adjustments tomorrow.

**Comment by vakhoir** [comment]
Just tested it, looks like your PR fixed the problem, but I created another #4843) one to clean up the registration, just to be sure.

**Comment by Web-eWorks** [comment]
Manually closing as fixed in #4842.
